src/sentry/snuba/metrics/datasource.py

```diff
@@ -420,6 +421,138 @@ def get_tag_values(
+    """Fields and values to filter queries when exceeding the Snuba query limit.
+
+    Snuba imposes a limit on the number of rows that can be queried and
+    returned. This limit can be exceeded when grouping metrics by one or more
+    tags. In this case, we take the first groups returned by Snuba and filter
+    subsequent queries with this set of tag values.
+
+    Fields:
+
+    - ``keys``: A tuple containing resolved tag names ("tag[123]") in the order
+      of the ``groupBy`` clause.
+
+    - ``values``: A list of tuples containing the tag values of the group keys.
+      The list is in the order returned by Snuba. The tuple elements are ordered
+      like ``keys``.
+
+    - ``conditions``: A list of raw snuba query conditions to filter subsequent
+      queries by.
+    """
+
+    keys: Tuple[Groupable]
+    values: List[Tuple[int]]
+    conditions: ConditionGroup
+
+
+def _get_group_limit_filters(
+    query: QueryDefinition, results: List[Mapping[str, int]]
+) -> Optional[GroupLimitFilters]:
+    if not query.groupby or not results:
+        return None
+
+    # Translate the groupby fields of the query into their tag keys because these fields
+    # will be used to filter down and order the results of the 2nd query.
+    # For example, (project_id, transaction) is translated to (project_id, tags[3])
+    keys = tuple(
+        resolve_tag_key(query.org_id, field) if field not in ALLOWED_GROUPBY_COLUMNS else field
+        for field in query.groupby
+    )
+
+    values = [tuple(row[col] for col in keys) for row in results]
+    conditions = [
+        Condition(Function("tuple", [Column(k) for k in keys]), Op.IN, Function("tuple", values))
+    ]
+
+    # In addition to filtering down on the tuple combination of the fields in
+    # the group by columns, we need a separate condition for each of the columns
+    # in the group by with their respective values so Clickhouse can filter the
+    # results down before checking for the group by column combinations.
+    values_by_column = {col: list({row[col] for row in results}) for col in keys}
+    conditions += [
+        Condition(Column(col), Op.IN, Function("tuple", col_values))
+        for col, col_values in values_by_column.items()
+    ]
+
+    return GroupLimitFilters(keys=keys, values=values, conditions=conditions)
+
+
+def _apply_group_limit_filters(query: Query, filters: GroupLimitFilters) -> Query:
+    where = list(filters.conditions)
+
+    for condition in query.where or []:
+        # If query is grouped by project_id, then we should remove the original
+        # condition project_id cause it might be more relaxed than the project_id
+        # condition in the second query. This does not improve performance, but the
+        # readability of the query.
+        if not (
+            isinstance(condition, Condition)
+            and isinstance(condition.lhs, Column)
+            and condition.lhs.name == "project_id"
+            and "project_id" in filters.keys
```

Comment:
This could check every key rather than just `project_id`. Would that make more sense?

Reply:
If this does not improve performance and only simplifies the snuba query, I would personally remove this special casing entirely. IMO simplicity of code is more important than simplicity of the constructed query.

Reply:
I get your point. I am not sure I agree though. If you have code that is generating incorrect queries, I would opt in to fix the code rather than having one less albeit a long composite condition in this case

Reply:
Fine with me!